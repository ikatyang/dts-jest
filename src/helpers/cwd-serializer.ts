export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
