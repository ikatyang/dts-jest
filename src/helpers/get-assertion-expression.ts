import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
