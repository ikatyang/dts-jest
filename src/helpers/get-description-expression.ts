import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
