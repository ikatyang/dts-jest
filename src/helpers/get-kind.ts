import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
