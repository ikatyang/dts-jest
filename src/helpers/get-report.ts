import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
