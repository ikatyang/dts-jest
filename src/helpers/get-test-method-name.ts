import {TriggerKind} from '../definitions';

export const get_test_method_name = (kind: TriggerKind) => {
  switch (kind) {
    case TriggerKind.None:
    case TriggerKind.Show:
      return 'test';
    case TriggerKind.Only:
      return 'test.only';
    case TriggerKind.Skip:
      return 'test.skip';
    default:
      throw new Error(`Unexpected kind ${JSON.stringify(kind)}`);
  }
};
