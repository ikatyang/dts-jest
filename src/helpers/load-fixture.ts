import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
