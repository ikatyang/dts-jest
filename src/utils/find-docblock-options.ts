import * as _ts from 'typescript';
import {
  docblock_option_regex,
  DocblockOptions,
  DocblockOptionMatchIndex,
  DocblockOptionValue,
  NormalizedConfig,
} from '../definitions';
import { for_each_comment } from './for-each-comment';

export const find_docblock_options = (
  source_file: _ts.SourceFile,
  ts: typeof _ts,
) => {
  const options: DocblockOptions = {};

  for_each_comment(
    source_file,
    comment => {
      if (!comment.startsWith('/**')) {
        return false;
      }

      const matched = comment.match(docblock_option_regex);

      if (matched === null) {
        return false;
      }

      const options_literal = matched[DocblockOptionMatchIndex.Options].trim();

      options_literal.split(/\s+/).forEach(option_value => {
        switch (option_value) {
          case DocblockOptionValue.EnableTestType:
            options.test_type = true;
            break;
          case DocblockOptionValue.EnableTestValue:
            options.test_value = true;
            break;
          case DocblockOptionValue.DisableTestType:
            options.test_type = false;
            break;
          case DocblockOptionValue.DisableTestValue:
            options.test_value = false;
            break;
          default:
            throw new Error(`Unexpected option value '${option_value}'`);
        }
      });

      return false;
    },
    ts,
  );

  return options;
};
