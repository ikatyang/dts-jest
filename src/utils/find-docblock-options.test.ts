import * as ts from 'typescript';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { find_docblock_options } from './find-docblock-options';

it('should not read from normal comment', () => {
  expect(find_options('normal-comment')).toEqual({});
});

it('should not read from non-first docblock', () => {
  expect(find_options('second-docblock')).toEqual({});
});

it('should find enable:test-type correctly', () => {
  expect(find_options('enable-test-type')).toMatchSnapshot();
});

it('should find enable:test-value correctly', () => {
  expect(find_options('enable-test-value')).toMatchSnapshot();
});

it('should find disable:test-type correctly', () => {
  expect(find_options('disable-test-type')).toMatchSnapshot();
});

it('should find disable:test-value correctly', () => {
  expect(find_options('disable-test-value')).toMatchSnapshot();
});

it('should find multi option value correctly', () => {
  expect(find_options('multi-option-value')).toMatchSnapshot();
});

it('should throw error if there is unexpected value', () => {
  expect(() => find_options('unexpected')).toThrowErrorMatchingSnapshot();
});

function find_options(id: string) {
  const source_file = load_fixture_source_file(
    `find-docblock-options/${id}.ts`,
    ts,
  );
  return find_docblock_options(source_file, ts);
}
