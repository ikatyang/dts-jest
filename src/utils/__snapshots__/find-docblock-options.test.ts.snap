// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should find disable:test-type correctly 1`] = `
Object {
  "test_type": false,
}
`;

exports[`should find disable:test-value correctly 1`] = `
Object {
  "test_value": false,
}
`;

exports[`should find enable:test-type correctly 1`] = `
Object {
  "test_type": true,
}
`;

exports[`should find enable:test-value correctly 1`] = `
Object {
  "test_value": true,
}
`;

exports[`should find multi option value correctly 1`] = `
Object {
  "test_type": false,
  "test_value": true,
}
`;

exports[`should throw error if there is unexpected value 1`] = `"Unexpected option value 'enable:nothing'"`;
